src/storage/src/hummock/compactor/mod.rs

```diff
@@ -225,6 +227,49 @@ impl Compactor {
             .await;
         let multi_filter_key_extractor = Arc::new(multi_filter_key_extractor);
 
+        let mut splits: Vec<KeyRange_vec> = vec![];
+        splits.push(KeyRange_vec::new(vec![], vec![]));
+        // collect sstable_ids to get meta data
+
+        let sstable_ids = select_table_infos
+            .iter()
+            .map(|sstable_info| sstable_info.id)
+            .collect_vec();
+        let mut indexes = vec![];
+
+        // preload the meta and get the smallest key to split sub_compaction
+        for sstable_id in sstable_ids {
+            // extend
+            indexes.extend(
+                context
+                    .sstable_store
+                    .sstable(sstable_id, &mut StoreLocalStatistic::default())
+                    .await
+                    .unwrap()
+                    .value()
+                    .meta
+                    .block_metas
+                    .iter()
+                    .map(|block| {
+                        FullKey::from_user_key_slice(user_key(&block.smallest_key), u64::MAX)
+                            .into_inner()
+                    })
+                    .collect_vec(),
+            );
+        }
+        indexes.sort();
+        // every 8 index as one group
+        const SPLIT_RANGE_STEP: usize = 8;
+        let concurrency = indexes.len() / SPLIT_RANGE_STEP;
+        let step = (indexes.len() + concurrency - 1) / concurrency;
+        for (idx, key) in indexes.into_iter().enumerate() {
+            if idx > 0 && idx % step == 0 {
+                splits.last_mut().unwrap().right = key.clone();
+                splits.push(KeyRange_vec::new(key, vec![]));
+            }
+        }
+
+        compact_task.splits = splits;
         // Number of splits (key ranges) is equal to number of compaction tasks
         let parallelism = compact_task.splits.len();
```

Comment:
1. concurrency
2. compact_task.splits.len()
3. parallelism

These concepts seem to be closed, hopefully we can reduce it in this pr. And `compact` function, is more than 100 line, we can seperate the implementation of leveraging the compact_task splits